### DISCUSSION TITLE
Move data from tres files to real data files

### DISCUSSION THREAD

**Discussion by berarma** [godotengine/godot-proposals#6871] [Engine Core]
Running "git diff" is usually painful due to `tres` files being text files sometimes holding lots of embedded data. Storing data with a string representation is inefficient, it uses more space that it's needed, and it makes working with these text files a bit harder.

I think data should be stored in binary data files and referenced from `tres` files. This would reduce the size of the project and working with `tres` files as text files would be easier. Loading them from Godot would probably be faster too since it wouldn't have to parse all those numbers and convert them to real data.

What do you think?

**Comment by AThousandShips** [comment]
The fact that `tres` files are text based is to make version control work smoother, so this goes against that core idea, you can make the active decision to do this by just storing resources and things in `res` files, most data in `tscn` files that is clunky can be stored in this way, the one exception I can think of is `TileMap`

**Reply by Zireael07** [reply]
> I think data should be stored in binary data files and referenced from tres file 

I think they are literally saying the same thing you do

> by just storing resources and things in res files

Anyway tres is text based to make version control work better. If you have lots of embedded data then it's usually user error, but that (and best practices such as what AThousandShips said) should be better emphasized in docs.

**Reply by AThousandShips** [reply]
Essentially: The general suggestion for data that is human-readable is to use text formats in version control, this is why there's the option to convert `tres` and `tscn` files to `res` and `scn` files on export for example

If you find it more useful to you to use binary formats then that's a choice you can make, but I don't think doing a hybrid choice is really that helpful

(didn't see your comment before posting hence saying some of the same things lol)

**Reply by berarma** [reply]
> The fact that `tres` files are text based is to make version control work smoother, so this goes against that core idea, you can make the active decision to do this by just storing resources and things in `res` files, most data in `tscn` files that is clunky can be stored in this way, the one exception I can think of is `TileMap`

Could you please point me to some explanation about how to do this? I'm still new to Godot and I haven't seen any option to store data in `res` and `scn` files. Godot automatically puts the data there.

**Reply by AThousandShips** [reply]
Meant resources, not specifically data in a scene sorry for not being more clear on that, to store the resources you need to save them as a file

**Comment by AThousandShips** [comment]
Also how should this handle if the file storing the binary data is deleted? What limits should be put to make the decision to store it one way and not another? Just some thoughts on the questions raised by this

**Reply by berarma** [reply]
I guess it would be the same as with any resource referencing another resource. It isn't anything new.

When to store it as data? Always when it's binary data being converted into a string of numbers.

As an example, I have this `tres` file made from a ttf font file:
[gd_resource type="FontFile" load_steps=3 format=3 uid="uid://c0wy22upupcbi"]

```[sub_resource type="Image" id="Image_5r238"]
data = {
"data": PackedByteArray(255, 0, 255, 0, 255,...
```

It goes on a on. I'm looking for some way to store the data in a binary file but I can't find it. Is even that data required since the ttf file is stored in the project?

**Reply by AThousandShips** [reply]
Not really, it wouldn't be a separate resource but a second file containing the same resource

**Reply by AThousandShips** [reply]
Why not just store it as a `res`?

**Reply by berarma** [reply]
> Why not just store it as a `res`?

So it's all or nothing. I meant keeping resources as text files except the embedded data.

I didn't know I could store as `res` and it works the same as a binary file. Thanks for that piece of information.

I'll try using that and see how it goes. But I still see no good reason to embed data in `tres` files. If the reason to have them is being more friendly to version control systems, they're no longer friendly with those long data strings.

**Reply by AThousandShips** [reply]
Yes, as I said above `res` are binary files

But human readable files are still the preferred way, and a hybrid format would create a lot of issues, add additional files that create clutter, create additional issues of accidentally moving or removing files, etc.

Do you have any examples of file formats that do this in other programs or game engines?

**Reply by AThousandShips** [reply]
When a resource file is deleted it causes any reference to it to fail, but there's a clear case when *all* the data vanishes, there's also tracking of the file moving managed, having a file that just holds the binary data would complicate this, how would it be named? How would it be tracked? What happens if you move just the base file and not the data?

**Reply by Zireael07** [reply]
> But I still see no good reason to embed data in tres files. 

Again. Version control can track changes to this data unlike with binary files, where it logs that the whole file was changed.

> they're no longer friendly with those long data strings.

Most of the data that *is* supposed to be embedded does NOT have long data string (eg shaders, materials, tilesets data). If you do, you most likely embedded something that really should not be, like a whole ArrayMesh or a Font file (those are very common mistakes to make)

Also: instead of downvoting the request to provide examples of hybrid data formats, please provide them. I admit I'm at a loss here as to how it would work and what it would look like.

**Reply by AThousandShips** [reply]
You say it's "all or nothing" but what is the actual issue with that? What situations do you have when you only want to be able to access part of the file in a human readable way? What parts of said file would be useful to be accessible that way?

**Reply by AThousandShips** [reply]
> When to store it as data? Always when it's binary data being converted into a string of numbers.

Even a file containing only a few numbers? Like a mesh with four vertices?

**Reply by berarma** [reply]
> You say it's "all or nothing" but what is the actual issue with that? What situations do you have when you only want to be able to access part of the file in a human readable way? What parts of said file would be useful to be accessible that way?

Anything in a PackedByteArray or similar.

**Reply by berarma** [reply]
> > But I still see no good reason to embed data in tres files.
> 
> Again. Version control can track changes to this data unlike with binary files, where it logs that the whole file was changed.
> 
> > they're no longer friendly with those long data strings.
> 
> Most of the data that _is_ supposed to be embedded does NOT have long data string (eg shaders, materials, tilesets data). If you do, you most likely embedded something that really should not be, like a whole ArrayMesh or a Font file (those are very common mistakes to make)
> 
> Also: instead of downvoting the request to provide examples of hybrid data formats, please provide them. I admit I'm at a loss here as to how it would work and what it would look like.

@AThousandShips seems pretty pissed even because this conversation is happening. I don't know which examples would make her change her very fond opinion. I understand some of her points but I don't think I've said something really that stupid,, although maybe.

**Reply by AThousandShips** [reply]
What are you replying to?

**Reply by AThousandShips** [reply]
Don't call me "he", and don't talk to me as if I'm not here, this is rude and completely improper

**Reply by berarma** [reply]
> Don't call me "he", and don't talk to me as if I'm not here, this is rude and completely improper

Sorry. I was just answering someone in the conversation and of course I'm replying as if you're here, of course. I'm out.

**Reply by AThousandShips** [reply]
Please take a look at the [code of conduct](https://godotengine.org/code-of-conduct/)

**Reply by AThousandShips** [reply]
You wrote "she seems..." that's talking about me as if I'm not here

**Reply by AThousandShips** [reply]
You still call me "he", basic courtesy to respect people's pronouns...

**Reply by AThousandShips** [reply]
I am providing polite, fair, and reasonable criticism and commentary, I am not "refusing to consider" this, I very calmly asked for examples of other programs using this concept, and I was not alone in thinking this was relevant

**Reply by AThousandShips** [reply]
> Always assume positive intent from others.

> Similarly, when other users offer you feedback please accept it gracefully.

> For certainty, any conduct which could reasonably be considered inappropriate in a professional setting is not acceptable.

**Reply by berarma** [reply]
> I am providing polite, fair, and reasonable criticism and commentary, I am not "refusing to consider" this, I very calmly asked for examples of other programs using this concept, and I was not alone in thinking this was relevant

Sorry. I talked on feelings. I shouldn't have to.

**Reply by AThousandShips** [reply]
Please take a look at the above referenced code of conduct before engaging further, this was a thoroughly unpleasant and offensive experience for me, I will leave now but do keep my suggestions and criticism in mind

**Comment by dalexeev** [comment]
I think that the main problem is long lines ([example](https://github.com/godotengine/godot-demo-projects/pull/861/files#diff-f39dbc0e584748909c23720ff76fe6e34f5b9d3639a06591d76b2af450cc97be)). There was a suggestion to wrap lines in tres/tscn after reaching some limit. The default context in git is 3 lines. If they are short, then this will not be a problem.
